### PR TITLE
Add process_cascade_events to the internal test tools

### PR DESCRIPTION
**Pull request by kitchoi**
Attribute errors such as https://github.com/enthought/traitsui/issues/854 are caused by events being processed after `UI.dispose` is called. Because various UI attributes are removed in `dispose`, the processing of these left-over events will cause errors.

This issue applies to assertions in the test body as well, which is why tests will typically call `GUI.process_events` prior to checking states. For more complex UIs, an event could easily cause a cascade of new events. For most testing purposes, we want to simulate the production environment where the GUI event loop is continuously running and we normally want all those cascade of events to be processed before we do assertions or dispose the UI.

The wx implementation of `pyface.gui.GUI.process_events` will process all cascade events, so it is enough to call `pyface.gui.GUI.process_events` just once before assertions and in between cleanup actions.

For Qt, the implementation of `pyface.gui.GUI.process_events` calls `QtCore.QCoreApplication.processEvents` which only processes the events that are in the queue when the function is called, excluding events that are posted by those events.

This PR introduces a function called `process_cascade_events` in the internal testing tools module so that we can process all cascaded events in Qt (and Wx). I will apply this function to all existing tests (in fact, I will open a draft PR that depends on this one to experiment the outcome with CI).

This function should probably be added to `pyface.gui.GUI`, once this is reviewed, I will open a separate PR to `pyface`, but it will be a feature there, rather than an internal utility function.

### CODE REVIEW

File: traitsui/tests/_tools.py [hunk 1]
```diff
@@ -131,6 +133,37 @@ def filter_tests(test_suite, exclusion_pattern):
...
+        # if there are pending events (`hasPendingEvents` is deprecated).
+        # The minus-one is to account for precision differences between Python
+        # and Qt, false positive triggers another redundant run that should
+        # return immediately so that is fine.
+        while start is None or duration_millisecs >= _TOLERANCE_MILLISECS - 1:
+            start = time.time()
+            QtCore.QCoreApplication.processEvents(
+                QtCore.QEventLoop.AllEvents, _TOLERANCE_MILLISECS
+            )
+            duration_millisecs = (time.time() - start) * 1000
```
Comment (kitchoi) [inline traitsui/tests/_tools.py:162]: Yep this is ugly and I'd be very happy if that is a neater solution.

Comment (kitchoi) [reply traitsui/tests/_tools.py:162]: There is `flush` but Qt documentation says it is deprecated. https://doc.qt.io/qt-5/qcoreapplication-obsolete.html#flush

File: traitsui/tests/test_tools.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        self.assertEqual(q_object.n_events, max_n_events)
+
+    @skip_if_not_wx
+    def test_wx_process_events_process_all(self):
+
+        def cleanup(wx_handler):
+            GUI.process_events()
+            wx_handler.Destroy()
+
+        wx_handler = DummyWxHandler(max_n_events=10)
```
Comment (ievacerny) [inline traitsui/tests/test_tools.py:122]: Nit: could we have a local variable `max_n_events` as in `test_qt_process_events_process_all`?

File: traitsui/tests/test_tools.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+        # then
+        self.assertEqual(q_object.n_events, max_n_events)
+
+    @skip_if_not_wx
+    def test_wx_process_events_process_all(self):
+
+        def cleanup(wx_handler):
+            GUI.process_events()
+            wx_handler.Destroy()
```
Comment (ievacerny) [inline traitsui/tests/test_tools.py:120]: Is the `GUI.process_events()` here in case of a test failure?

File: traitsui/tests/_tools.py [hunk 4]
```diff
@@ -131,6 +133,37 @@ def filter_tests(test_suite, exclusion_pattern):
...
+        from pyface.qt import QtCore
+        start = None
+
+        # Qt won't raise if there are still events to be processed before
+        # the time limit is reached. There are no other safe way to tell
+        # if there are pending events (`hasPendingEvents` is deprecated).
+        # The minus-one is to account for precision differences between Python
+        # and Qt, false positive triggers another redundant run that should
+        # return immediately so that is fine.
+        while start is None or duration_millisecs >= _TOLERANCE_MILLISECS - 1:
```
Comment (ievacerny) [inline traitsui/tests/_tools.py:157]: 🤖 flake8 is complaining about undefined variable `duration_millisecs `

File: traitsui/tests/test_tools.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
+#  Copyright (c) 2008-20, Enthought, Inc.
```
Comment (ievacerny) [inline traitsui/tests/test_tools.py:1]: Nit: shouldn't this file be named `test__tools.py`?

File: traitsui/tests/test_tools.py (hunk 3 above)
Comment (kitchoi) [reply traitsui/tests/test_tools.py:120]: Yep. I will add a comment here to explain that.

**Pull request merged**